### ISSUE TITLE
Attributes Form Editor - Can't have multiple instances of same field

### ISSUE THREAD

**Issue by Saijin-Naib** [qgis/QGIS#33068]
**Describe the bug**
If an attribute form has multiple widgets pointing to the same field, they will all share the same value and will reflect one another when the value is changed.

**How to Reproduce**
Have an attributes form with multiple widgets pointed to the same field. Why would you have a form set up this way? Imagine you have a form where you want to track the Type, Extent, Severity for multiple different Distress types (Drainage, Roughness, Treatment Condition, Alligator Tracking, Long/Trans Cracking, Patching/Potholes, Rutting, Bleeding/Raveling) for a given type of road cover. The way the target database is modeled, each type of Distress gets a unique numerical ID and Extent and Severity both get L/M/H values that get summed and stored in a Distress_Value column.

With this setup, the Attributes Form as they stand can not be used. By setting the Extent or Severity in one Distress, you set it across ALL distress values. Despite being present multiple different times, each widget is effectively the same "instance" for writing out to the dataset.

**QGIS and OS versions**

QGIS version | 3.10.0-A Coruña | QGIS code revision | 6ffa89eb3e
-- | -- | -- | --
Compiled against Qt | 5.11.2 | Running against Qt | 5.11.2
Compiled against GDAL/OGR | 3.0.2 | Running against GDAL/OGR | 3.0.2
Compiled against GEOS | 3.8.0-CAPI-1.13.1 | Running against GEOS | 3.8.0-CAPI-1.13.1
Compiled against SQLite | 3.29.0 | Running against SQLite | 3.29.0
PostgreSQL Client Version | 11.5 | SpatiaLite Version | 4.3.0
QWT Version | 6.1.3 | QScintilla2 Version | 2.10.8
Compiled against PROJ | 6.2.1 | Running against PROJ | Rel. 6.2.1, November 1st, 2019
OS Version | Windows 10 (10.0)
Active python plugins | CalculateGeometry; cartogram3; changeDataSource; DataPlotly; DEMto3D; dzetsaka; GeoCoding; GeodesicDensifier; geomovetools; getools; GroupPointsWithinDistance; guided_offline_editing; htpgeoprocessor; joinmultiplelines; karika; kmltools; kml_style_importer; LandSurveyCodesImport; LAStools; latlontools; LineDirectionHistogram; lrs; maps2winbugs; MemoryLayerSaver; minimum_spanning_tree; mmqgis; networks; openlayers_plugin; oscars-georaster-qgis-plugin; ParallelLineConstruction; pdaltools; photo2shape; PluginLoadTimes; Polygons_Pile; polystrip; processing_circuitscape; processing_fusion; processing_gnm; processing_js; processing_pktools; processing_prepair; processing_r; processing_saga_nextgen; processing_taudem; processing_trajectory; processing_whitebox; project_colors_dock; qchainage; qconsolidate; qdraw; qgiscloud; QGISSortAndNumber-master; QRectangleCreator; QuickOSM; realcentroid; refFunctions; SelectWithin; sen2cor_adapter; Serval; shapetools; snail; spatial_query_with_values; SplitPolygonShowingAreas; SpreadsheetLayers; statist; ThToolBox; toggle_labels_widget; trail_elevation_stats; wnt; db_manager; processing

**Additional context**
Attempting to make a Attributes Form to interact with the CAMP-RS database.
Here's a sample form/print layout I made in QGIS for collecting, I want to use the Attributes Form for future collecting as well as current data entry.
[Road Survey Form - Optimized 1.pdf](https://github.com/qgis/QGIS/files/3888998/Road.Survey.Form.-.Optimized.1.pdf)

Here's how CAMP-RS's own tool exposes the survey interface:  
![image](https://user-images.githubusercontent.com/19295950/69582922-a75bfc80-0fa7-11ea-93bb-05818ab9db7d.png)

Here's how that data gets written to the database:  
![image](https://user-images.githubusercontent.com/19295950/69582982-c5c1f800-0fa7-11ea-8d43-bbfae7b80c50.png)

**Comment by gioman** [comment]
> The way the target database is modeled, each type of Distress gets a unique numerical ID and Extent and Severity both get L/M/H values that get summed and stored in a Distress_Value column.

So you are entering multiple records, and that is not how works in QGIS (and in GIS in general?). One form is to enter the attributes (1 record) for 1 geometry. At best this is a feature request but this could be even better suited for a plugin rather than a core functionality.

**Comment by Saijin-Naib** [comment]
I don't think this is quite that exotic, to be honest. This is just a One-To-Many relationship, and again, with a properly normalized database, this type of relationship would be quite common, especially if you are tracking historical information for a given feature(s).

For instance, another asset within our town we have to keep track of is our stormwater collection ponds. These are fairly stable geometries in that once they exist, they pretty much stay exactly as they were. However, we are bound by regulation to inspect each one at a minimum once every three years. We store those inspections on a table, one entry for each inspection. That table is related to the geometries by a unique ID for each pond. Using the QGIS Processing Algorithm _Join Attributes by Field Value_, I can then export the aggregated data for usage in other applications.

**Comment by gioman** [comment]
> I don't think this is quite that exotic, to be honest. This is just a One-To-Many relationship, and again, with a properly normalized database, this type of relationship would be quite common, especially if you are tracking historical information for a given feature(s).

@Saijin-Naib 
you already have 1:n relations in QGIS where you can edit both sides of the join in "single" form. Does not fit your needs?



> For instance, another asset within our town

have you considered supporting the work needed to implement exactly what you need?

**Comment by Saijin-Naib** [comment]
> @Saijin-Naib
> you already have 1:n relations in QGIS where you can edit both sides of the join in "single" form. Does not fit your needs?

No, I've not been able to make that work the way I'm looking to. It is entirely possible that I'm just not implementing it correctly on my end, however.

> have you considered supporting the work needed to implement exactly what you need?

I have indeed, and I've had this discussion a few times with Nyall already.
In summary: 

- I'm acting as an independent end user/QGIS fan. 
- I lack the personal resources to fund the project more than small (sub-$100USD) donations sporadically. 
- I lack any real/appreciable programming skill. Look at my GitHub commit history if you want a laugh.
- I do have access to some larger GIS organizations and people in my network, which I do evangelize QGIS and related projects to. 
- I lack the personal capital at my employment to have gotten them to invest/fund QGIS development, despite my consistent and hard pushes to move us away from Arc. 
- When I can secure funding, I will 100% do so. 

Until then, I'm just a low-resource end user. Unless something like this is bite-sized enough for me to fund, or someone is flexible enough to setup a lot of small "gates" that I can fund as my budget allows.

**Comment by elpaso** [comment]
@Saijin-Naib your inputs are valuable, we definitely need use cases to define out feature road map.

For the funding, you might consider writing down a QEP and launching a crowdfunding campaign, there have been quite a few successful campaigns in the past.

**Comment by gioman** [comment]
> No, I've not been able to make that work the way I'm looking to. It is entirely possible that I'm just not implementing it correctly on my end, however.

@Saijin-Naib let us know if you are able to. Anyway my understanding of your issue/request description is that is something slightly different than QGIS 1:n relations --> while editing one table you may want to add multiple records (that implies replicating n times the digitized geometry) where a few attributes values are equal and where other attributes differ. The attributes that different coming from repeated entries in the input form. Is that correct?

> another asset within our town

as you wrote "our town" I assumed you were speaking in name of a city council or similar. Usually such entities have plenty of funds to buy software licenses.

**Comment by Saijin-Naib** [comment]
> @Saijin-Naib your inputs are valuable, we definitely need use cases to define out feature road map.
> For the funding, you might consider writing down a QEP and launching a crowdfunding campaign, there have been quite a few successful campaigns in the past.

Thanks, @elpaso. I try my best to at least contribute meaningful suggestions.

Thank you for the suggestion. I'm trying to work on a funding campaign already with @nyalldawson . Pending how well I manage this one, I'll consider starting further campaigns. I'm not entirely confident that I'll be able to drive a campaign through to success.

> @Saijin-Naib let us know if you are able to. Anyway my understanding of your issue/request description is that is something slightly different than QGIS 1:n relations --> while editing one table you may want to add multiple records (that implies replicating n times the digitized geometry) where a few attributes values are equal and where other attributes differ. The attributes that different coming from repeated entries in the input form. Is that correct?

@gioman , not necessarily duplicating the geometry. Quite the opposite actually. I intend fully for the geometry table/records to be unique, and for the attributes table/records to hold N-number of records associated with each feature, where appropriate.

As for the attribute form interaction, yes, that would be the case. A constant in each record would likely be RIN (geohash), QGIS username variable, QGIS current date. If we take the case submitted above, you may have multiple entries for a given road segment (geohash) with Distress_IDs and Distress_Values differing. These represent numerical assessments (Low/Medium/High) of different pavement distress categories. If you look at the attribute table in this way, it serves to store each survey criteria & value in a list which can be many-to-one joined to a given road segment via the RIN (geohash). This makes sense since these records all describe a single geometry that does not change, but is instead being observed multiple times.

> as you wrote "our town" I assumed you were speaking in name of a city council or similar. Usually such entities have plenty of funds to buy software licenses.

Normally, yes. Likely irrelevant here, but my position is not necessarily permanent, pending whether or not the town board determines that I've given them enough value for the funds they've diverted from other areas to form my salary. Along those lines, software budget is inflexible as it covers specifically only a low-level Arc license, and without moving them away from that, there are no funds available to me to fund development. I have yet to convince them to move from Arc due to (perceived) shortcomings of the OSGEO stack compared to Arc. Chicken and egg problem.

Thank you all for your help, by the way.

**Comment by gioman** [comment]
> @gioman , not necessarily duplicating the geometry. Quite the opposite actually. I intend fully for the geometry table/records to be unique, and for the attributes table/records to hold N-number of records associated with each feature, where appropriate.
> 
> As for the attribute form interaction, yes, that would be the case. A constant in each record would likely be RIN (geohash), QGIS username variable, QGIS current date. If we take the case submitted above, you may have multiple entries for a given road segment (geohash) with Distress_IDs and Distress_Values differing. These represent numerical assessments (Low/Medium/High) of different pavement distress categories. If you look at the attribute table in this way, it serves to store each survey criteria & value in a list which can be many-to-one joined to a given road segment via the RIN (geohash). This makes sense since these records all describe a single geometry that does not change, but is instead being observed multiple times.

@Saijin-Naib so... it is really a case for the relations (1:n joins) in QGIS, but probably with data entry done in different way compared to what we have now. So this is a feature request, not a bug. And I also agree with Alessandro when he says that probably this is a better fit for a QEP.
